### PR TITLE
set the thumbprint to enable secure communications

### PR DESCRIPTION
**Pull request by yastij**
Signed-off-by: Yassine TIJANI <ytijani@vmware.com>

**What this PR does / why we need it**: This PR adds Thumbprint support and plumb it through the `client`, This should enable secure communications

**Which issue(s) this PR fixes**: Fixes https://github.com/kubernetes-sigs/cluster-api-provider-vsphere/issues/1051

**Special notes for your reviewer**:

**Release note**:

```release-note
Thumbprint support is added to the clients used, this enable secure communication between CAPV and vcenter
```

### CODE REVIEW

**Comment by yastij** [comment]
/hold

holding for thumbprint support in https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/499

**Comment by yastij** [comment]
this should be good to go, we can add the csi bits later once it gets released

File: api/v1alpha3/types.go [hunk 1]
```diff
@@ -78,6 +78,11 @@ type VirtualMachineCloneSpec struct {
 	// +optional
 	Server string `json:"server,omitempty"`
 
+	// Thumbprint is the thumbprint of the given Server
+	// when supplied Insecure should not be set to true
```
Comment (fabriziopandini) [inline api/v1alpha3/types.go:82]: This description should be adapted to the machine object (e.g. there is no insecure at the machine level)

File: api/v1alpha3/vspherecluster_types.go [hunk 2]
```diff
@@ -39,6 +39,11 @@ type VSphereClusterSpec struct {
 	// +optional
 	Insecure *bool `json:"insecure,omitempty"`
 
+	// Thumbprint is the thumbprint of the given Server
+	// when supplied Insecure should not be set to true
+	// +optional
+	Thumbprint string `json:"thumbprint,omitempty"`
```
Comment (fabriziopandini) [inline api/v1alpha3/vspherecluster_types.go:45]: Question:
Given that now `insecure==true` is equivalent to `Thumbprint == ""`, is it required to preserve the `insecure` flag? If not removing, at least what about deprecating?

**Review by fabriziopandini — COMMENTED** [review]
overall lgtm

File: api/v1alpha3/vspherecluster_types.go (hunk 2 above)
Comment (yastij) [reply api/v1alpha3/vspherecluster_types.go:45]: `insecure` is planned for deprecation as per our v1alpha4 roadmap. I'm planning on sending a different PR for this

File: api/v1alpha3/types.go (hunk 1 above)
Comment (yastij) [reply api/v1alpha3/types.go:82]: ack

File: api/v1alpha3/types.go [hunk 3]
```diff
@@ -78,6 +78,12 @@ type VirtualMachineCloneSpec struct {
 	// +optional
 	Server string `json:"server,omitempty"`
 
+	// Thumbprint is the thumbprint of the given Server
+	// when this is set to empty, this VirtualMachine would be created
+	// with an insecure communication
+	// +optional
```
Comment (randomvariable) [inline api/v1alpha3/types.go:84]: Nit
```suggestion
	// Thumbprint is the colon-separated SHA-1 checksum of the given vCenter server's host certificate
	// When this is set to empty, this VirtualMachine would be created
	// without TLS certificate validation of the communication between Cluster API Provider vSphere
	// and the VMware vCenter server.
	// +optional
```

File: api/v1alpha3/vspherecluster_types.go [hunk 4]
```diff
@@ -39,6 +39,11 @@ type VSphereClusterSpec struct {
 	// +optional
 	Insecure *bool `json:"insecure,omitempty"`
 
+	// Thumbprint is the thumbprint of the given Server
+	// when supplied Insecure should not be set to true
+	// +optional
```
Comment (randomvariable) [inline api/v1alpha3/vspherecluster_types.go:44]: Nit
```suggestion
	// Thumbprint is the colon-separated SHA-1 checksum of the given vCenter server's host certificate
	// When provided, Insecure should not be set to true
	// +optional
```

File: packaging/flavorgen/flavors/generators.go [hunk 5]
```diff
@@ -44,6 +44,7 @@ const (
 	machineDeploymentNameSuffix  = "-md-0"
 	namespaceVar                 = "${NAMESPACE}"
 	vSphereDataCenterVar         = "${VSPHERE_DATACENTER}"
+	vSphereTumbprint             = "${VSPHERE_SSL_THUMBPRINT}"
```
Comment (randomvariable) [inline packaging/flavorgen/flavors/generators.go:47]: nit
```suggestion
	vSphereTumbprint             = "${VSPHERE_TLS_THUMBPRINT}"
```

File: packaging/flavorgen/flavors/generators.go [hunk 6]
```diff
@@ -121,15 +122,19 @@ func newVSphereCluster(lb *infrav1.HAProxyLoadBalancer) infrav1.VSphereCluster {
 			Namespace: namespaceVar,
 		},
 		Spec: infrav1.VSphereClusterSpec{
-			Server: vSphereServerVar,
+			Server:     vSphereServerVar,
+			Thumbprint: vSphereTumbprint,
```
Comment (randomvariable) [inline packaging/flavorgen/flavors/generators.go:126]:
```suggestion
			Thumbprint: vSphereThumbprint,
```

File: packaging/flavorgen/flavors/generators.go [hunk 7]
```diff
@@ -121,15 +122,19 @@ func newVSphereCluster(lb *infrav1.HAProxyLoadBalancer) infrav1.VSphereCluster {
...
 			CloudProviderConfiguration: infrav1.CPIConfig{
 				Global: infrav1.CPIGlobalConfig{
 					SecretName:      "cloud-provider-vsphere-credentials",
 					SecretNamespace: metav1.NamespaceSystem,
-					Insecure:        true,
+					Thumbprint:      vSphereTumbprint,
```
Comment (randomvariable) [inline packaging/flavorgen/flavors/generators.go:131]:
```suggestion
					Thumbprint:      vSphereThumbprint,
```

File: packaging/flavorgen/flavors/generators.go [hunk 8]
```diff
@@ -121,15 +122,19 @@ func newVSphereCluster(lb *infrav1.HAProxyLoadBalancer) infrav1.VSphereCluster {
...
 				},
 				VCenter: map[string]infrav1.CPIVCenterConfig{
-					vSphereServerVar: {Datacenters: vSphereDataCenterVar},
+					vSphereServerVar: {
+						Datacenters: vSphereDataCenterVar,
+						Thumbprint:  vSphereTumbprint,
```
Comment (randomvariable) [inline packaging/flavorgen/flavors/generators.go:136]:
```suggestion
						Thumbprint:  vSphereThumbprint,
```

File: test/e2e/config/vsphere-ci.yaml [hunk 9]
```diff
@@ -86,6 +86,7 @@ variables:
   EXP_CLUSTER_RESOURCE_SET: "true"
   CONTROL_PLANE_MACHINE_COUNT: 1
   WORKER_MACHINE_COUNT: 1
+  VSPHERE_SSL_THUMBPRINT: "4F:AF:22:BD:C8:B2:AC:DE:0A:04:71:02:D1:62:05:50:2B:A2:54:E9"
```
Comment (randomvariable) [inline test/e2e/config/vsphere-ci.yaml:89]:
```suggestion
  VSPHERE_TLS_THUMBPRINT: "4F:AF:22:BD:C8:B2:AC:DE:0A:04:71:02:D1:62:05:50:2B:A2:54:E9"
```

**Comment by randomvariable** [comment]
Mostly comments on the API documentation and otherwise typos, but logic otherwise lgtm

**Pull request merged**